examples/lin_algebra/gmres.cpp

```diff
@@ -0,0 +1,274 @@
+
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+#include <arrayfire.h>
+#include <af/util.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <vector>
+#include <iostream>
+using namespace af;
+
+struct info_solver {
+  std::vector<float> iter;
+  std::vector<float> time;
+  std::vector<float> nnz;
+};
+template <class T>
+void Update(array& x, const int k, const std::vector<T>&H,
+            const int R, const std::vector<T>&s,
+            const array v)
+{
+  T y[k+1];
+  for (int i=0; i<k+1; i++)
+    y[i] = s[i];
+  // Back substituation, 
+  // solve y, where H*y = s
+  for (int i = k; i >= 0; i--)
+  {
+    y[i] = y[i]/ H[i + i*(R+1)];
+    for (int j=i-1; j >= 0; j--)
+    {
+      y[j] = y[j] - H[j + i*(R+1)] * y[i];
+    }
+  }
+  // x = v*y
+  for (int j = 0; j <= k; j++){   
+    x+=v.col(j)*y[j];
+  }
+}
+template<class T>
+void ApplyPlaneRotation(T &dx, T &dy, T &cs, T &sn)
+{
+  T temp  =  cs * dx + sn * dy;
+  dy = -sn * dx + cs * dy;
+  dx = temp;
+}template<class T>
+void GeneratePlaneRotation(T &dx, T &dy, T &cs, T &sn)
+{
+  if (dy == 0.0) {
+    cs = 1.0;
+    sn = 0.0;
+  } else if (abs(dy) > abs(dx)) {
+    T temp = dx / dy;
+    sn = 1.0 / sqrt( 1.0 + temp*temp );
+    cs = temp * sn;
+  } else {
+    T temp = dy / dx;
+    cs = 1.0 / sqrt( 1.0 + temp*temp );
+    sn = temp * cs;
+  }
+}
+template<class T>
+int GMRES(array& x, const array& A, const array& b, const int n, const int R, const T tol, const int max_iter, dtype dt) {  
+  array w(n, dt);
+  std::vector<T> H((R+1)* R);
+  std::vector<T> s(R+1);
+  std::vector<T> cs(R+1);
+  std::vector<T> sn(R+1);
+  array v( n, R+1, dt);
+  array temp_arr(n,dt);
+  T normb = norm(b);
+  array r = b - matmul(A, x);
+  T beta = norm(r);
+  if (beta < tol) {
+    return 0;
+  }
+  int iter = 0;
+  while ( iter< max_iter )
+  {
+    v.col(0) = r / beta;
+    std::fill(s.begin(), s.end(), 0);
+    
+    s[0] = beta;
+    for ( int i=0; i< R; i++ )
+    {
+      iter++;
+      // XXLiu: w = solve(A * v[i]);
+      w = matmul(A, v.col(i));
+
+      int h_idx;
+      for (int k=0; k<=i; k++)
+      {
```

Comment:
Code could use some more comments here to explain what is going. Looks like we are doing Arnoldi iteration here. Looks like Givens rotations are being performed on the Hessenberg matrix.